Added staff and instructor roles on ccx to all the staff and instructors of the master course plus fixed view as student masquerade and display name of ccx on coach dashboard

amir-qayyum-khan opened the pull request:
### Background
fixes https://github.com/mitocw/edx-platform/issues/126, https://github.com/mitocw/edx-platform/issues/155 and https://github.com/mitocw/edx-platform/issues/115

### What is done in this PR
**Studio Updates:** 
- Remove ccx from staff dashboard inside studio. Because  a master course can have unlimited ccx. If we allow display of all ccx of logged in staff (who is also staff of master course ) in studio dashboard then his (staff's) dashboard can be populate with huge list of ccx along with master course 

**LMS Updates:** 
- Now when a coach creates a new ccx, I am assigning staff role on this ccx course to all staff of master course.
- Now staff of master course has staff rights on master course as well as ccx courses.
- This is for course staff to see what ccx coaches are doing with course.
- Added instructors of master course as instructor(s) in ccx.
- **Note:** This code will add/enrol staff and instructors in ccx and now max student limit for ccx depends on number of staff and admins. 
- Fixed view as student issue on ccx, now staff can view as student and see how ccx  view will appear to student
- Fixed display name on ccx coach dashboard. Previously it was show display name of master course on coach dashboard.

@pdpinch @giocalitri @pwilkins 

- Staff on Master course
![screen shot 2015-11-30 at 5 12 54 pm](https://cloud.githubusercontent.com/assets/10431250/11471231/ba5e9cc8-9785-11e5-95c5-0cdcea97f780.png)

- Staff on CCX
![screen shot 2015-11-30 at 5 13 04 pm](https://cloud.githubusercontent.com/assets/10431250/11471234/c2d0a9dc-9785-11e5-865f-a1b7d79a34db.png)

- Admin/instructors on master course
![screen shot 2015-12-10 at 5 10 18 pm](https://cloud.githubusercontent.com/assets/10431250/11715098/08685ed0-9f61-11e5-8b79-87defb82c2c9.png)

- Admin/instructors on ccx
![screen shot 2015-12-10 at 5 10 22 pm](https://cloud.githubusercontent.com/assets/10431250/11715099/11ada5ae-9f61-11e5-92c3-d0316f1b6190.png)

#### Fixed view as student

![screen shot 2015-12-17 at 7 21 37 pm 2](https://cloud.githubusercontent.com/assets/10431250/11871777/8eb3d64e-a4f3-11e5-9b32-d0df4a0dd0a6.png)
![screen shot 2015-12-17 at 7 21 53 pm 2](https://cloud.githubusercontent.com/assets/10431250/11871778/8eb92040-a4f3-11e5-9bb7-49c691d4bb24.png)
![screen shot 2015-12-17 at 7 22 05 pm 2](https://cloud.githubusercontent.com/assets/10431250/11871779/8ebdc94c-a4f3-11e5-9e7d-0ee44b1652c3.png)

#### Fixed display name of ccx on coach dashboard.
![screen shot 2015-12-18 at 3 29 34 pm](https://cloud.githubusercontent.com/assets/10431250/11894922/3f39bfa8-a59c-11e5-82e9-0dabd349e963.png)
![screen shot 2015-12-18 at 3 29 44 pm](https://cloud.githubusercontent.com/assets/10431250/11894921/3f3880de-a59c-11e5-9d38-da5de6837b8a.png)
![screen shot 2015-12-18 at 3 29 55 pm](https://cloud.githubusercontent.com/assets/10431250/11894923/3f411bf4-a59c-11e5-80bb-b79b5073c936.png)

edx-webhook commented:
Thanks for the pull request, @amir-qayyum-khan! I've created [OSPR-1007](https://openedx.atlassian.net/browse/OSPR-1007) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

pdpinch commented:
@amir-qayyum-khan as indicated in mitocw#126 the instructor/admin role needs to be applied to the CCX as well as the staff role. It looks like you will need to do that explicitly.

amir-qayyum-khan commented:
@pdpinch Done and rebased

amir-qayyum-khan commented:
jenkins run lettuce

pdpinch commented:
@amir-qayyum-khan I've found another issue with this. 

Giving master course staff access to the CCX as staff makes the CCX appear in their studio listing. Even worse, the overridden display-name isn't used. 

If there's an easy way to do it, I think we should suppress the display of CCXs in the studio listing. I think that would have to be a separate PR. Can you look into this?

An alternative would be to apply the display_name override, but that would still have the issue of polluting the studio listing with potentially hundreds of CCXs. 

I'm changing the title to WIP because I don't think this should be merged until this is resolved.

amir-qayyum-khan commented:
Done @pdpinch

amir-qayyum-khan commented:
jenkins run lettuce

amir-qayyum-khan commented:
jenkins run lettuce

pdpinch commented:
@ormsbee I wanted to run this by you. In particular, adding course staff to the CCX has the side affect of exposing the CCXs through Studio. We're taking the easy way out here & surpressing the listing of CCXs on the studio dashboard -- is that adequate (for now)? It does avoid the issue of giving the Davidson folks a list of 300 courses on studio.edge.edx.org

ormsbee commented:
@pdpinch: is the goal to actually make them recognized as staff for that course (e.g. so they can post on their forums and show with a "staff" label), or to just let them see/have permissions? For instance, if in the access checks, we did an extra thing that said, "if this is a CCX key and you're staff of the root course, you also have access", would that do the right thing for your use case?

To be clear, that's not my recommendation yet -- I'm just trying to figure out the goals.

ormsbee commented:
Also, just reading the description, it seems like this should be three separate PRs. Do they need to be bundled together?

pdpinch commented:
@ormsbee 

At the moment, the use case is being driven by support issues. We want to be able to see how the CCX is scheduled and assist with grading and student issues. In the medium term, I think instructors are curious to see how the MOOC is being used. At the same time, I didn't think we wanted to complicate all the permissions checks (e.g. if this is a CCX and if the user has admin then let them masquerade)

Regarding making it into 3 PRs, your processes encourage us to do fewer PRs, so we err on the side of lumping instead of splitting. @amir-qayyum-khan are there any internal dependencies, or can we break these into different PRs?

ormsbee commented:
@pdpinch: Yeah, permissions paths are a pain, but I'm also concerned about duplicated data and keeping things in sync (someone else takes over the course next term and is added, someone's removed, etc.). Would the staff of the MOOC eventually want to interact directly with students in the CCX via the forums (when that's working)?

My suggestion on splitting up the PR was in the hope that it would make it _easier_ to merge. If that hasn't been your experience, I won't insist on it.

amir-qayyum-khan commented:
@pdpinch @ormsbee should i divid this PR into 3 PR.
By they way I dont see any harm in merging this PR.

pdpinch commented:
@amir-qayyum-khan can you rebase this?

pdpinch commented:
@ormsbee w/r/t your question 

> Would the staff of the MOOC eventually want to interact directly with students in the CCX via the forums (when that's working)?

That's highly unlikely. MOOC staff see themselves as wholesalers and they don't want to get directly involved in the retail experience, especially discussions.

amir-qayyum-khan commented:
Done @pdpinch

pdpinch commented:
Removed WIP label

pdpinch commented:
@amir-qayyum-khan "View as Specific Student" is working for me now, but "View as Student" is not. It seems to be honoring the overridden dates, but it still displays sections that should be hidden from students.

amir-qayyum-khan commented:
@pdpinch I tested it, It is working fine for both as student and as specific student.
To hide chapter from student you must change start date of the chapter from schedule.

#### Tested scenarios:
- Changed start date of section to some next date from today.
- Changed start date of subsection to some next date from today.
- Changed due date of subsection to some next date from today.
- Changed start and due date of subsection from yesterday to next day.

All scenarios are working fine.

pdpinch commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -390,6 +391,13 @@ def course_filter(course):
+        # IF course is an instance of CCX and logged in user is the staff then do not show CCX(s) in his
+        # studio (aka cms) dashboard. Because for a single master course there can be many (no limit) CCX.
+        # Otherwise his dashboard can be fill with huge list of CCXs of multiple master course.
+        if isinstance(course, CCXLocator):
```
Let's simplify this comment:

```
# Custom Courses for edX (CCX) is an edX feature for re-using course content. 
# CCXs cannot be edited in Studio (aka cms) and should not be shown in this dashboard.
```

amir-qayyum-khan replied:
Fixed

pdpinch commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -434,8 +442,13 @@ def _accessible_courses_list_from_groups(request):
+            # IF course is an instance of CCX and logged in user is the staff then do not show CCX(s) in his
+            # studio (aka cms) dashboard. Because for a single master course there can be many (no limit) CCX.
+            # Otherwise his dashboard can be fill with huge list of CCXs of multiple master course.
+            if course is not None and not isinstance(course, ErrorDescriptor) and not isinstance(course.id, CCXLocator):
```
See edited comment above in https://github.com/edx/edx-platform/pull/10877/files#r49656974

amir-qayyum-khan replied:
fixed

pdpinch commented on lms/djangoapps/ccx/utils.py:
```diff
@@ -250,3 +251,45 @@ def assign_coach_role_to_ccx(ccx_locator, user, master_course_id):
+    list_staff = list_with_level(course, 'staff')
+    list_instructor = list_with_level(course, 'instructor')
+
+    # To avoid circular imports
```
is this comment still necessary?

amir-qayyum-khan replied:
Fixed

pdpinch commented:
Ok. I can't reproduce the bug I found in https://github.com/edx/edx-platform/pull/10877#issuecomment-168672085 

:+1: from me.

@giocalitri can you do a review as well? Can you also check that creating a CCX via REST API will add these roles as well?

amir-qayyum-khan commented:
@pdpinch Done with adding a data migration to add staff of master course to CCX for all existing ccx.

Thanks,

pdpinch commented:
@giocalitri can you take a look at the migration that Amir added? It's supposed to add the master courses's admins and staff to existing CCXs.

pdpinch commented:
@amir-qayyum-khan do you have any idea how long the migration might take for, say, 300 CCXs?

giocalitri commented on lms/djangoapps/ccx/migrations/0002_add_master_course_staff_in_ccx#10877.py:
```diff
@@ -0,0 +1,32 @@
+# -*- coding: utf-8 -*-
```
question: is putting the PR # in the migration name an edx practice?

amir-qayyum-khan replied:
No, I just put it by my self

giocalitri replied:
ok, so it is better to remove it

giocalitri commented on lms/djangoapps/ccx/migrations/0002_add_master_course_staff_in_ccx#10877.py:
```diff
@@ -0,0 +1,32 @@
+from courseware.courses import get_course_by_id
+
+from lms.djangoapps.ccx.utils import add_master_course_staff_to_ccx
+from lms.djangoapps.ccx.models import CustomCourseForEdX
```
Nit: sorting of the imports is wrong

giocalitri commented on lms/djangoapps/ccx/migrations/0002_add_master_course_staff_in_ccx#10877.py:
```diff
@@ -0,0 +1,32 @@
+    """
+    list_ccx = CustomCourseForEdX.objects.all()
+    for ccx in list_ccx:
+        course = get_course_by_id(ccx.course_id)
```
I don't think you need to do this: the CCX model has a `course` method that should return the course

giocalitri commented:
@pdpinch the migration seems to be fine.

Semi-related question: I suppose I need to do the same operation (add master course staff to CCX) when the CCX is created through REST APIs, right?

pdpinch commented:
Yes, please!

> I suppose I need to do the same operation (add master course staff to CCX) when the CCX is created through REST APIs, right?

amir-qayyum-khan commented:
Done @giocalitri 

cc @pdpinch

amir-qayyum-khan commented:
@pdpinch 
> @amir-qayyum-khan do you have any idea how long the migration might take for, say, 300 CCXs?

No idea it depends the email service (each user will inform via email) and server hardware.

giocalitri commented on lms/djangoapps/ccx/migrations/0002_add_master_course_staff_in_ccx.py:
```diff
@@ -0,0 +1,29 @@
+    list_ccx = CustomCourseForEdX.objects.all()
+    for ccx in list_ccx:
+        ccx_locator = CCXLocator.from_course_locator(ccx.course_id, unicode(ccx.id))
+        add_master_course_staff_to_ccx(ccx.course, ccx_locator, ccx.display_name)
```
question: did you test that using `ccx.course` actually works?

amir-qayyum-khan replied:
yes , I tested it by re-runing migration. 

````
- python manage.py lms migrate ccx 0001  --fake --settings=devstack 
- python manage.py lms migrate ccx 0002 --settings=devstack 
````

pdpinch replied:
Can the migration be reversed?

I'm just remembering an unfortunate PR that we had to revert and it was further complicated because we couldn't revert the migrations. 

@bdero may have some advice. 

- Peter

amir-qayyum-khan commented:
Done with adding reverse migration 

@pdpinch

giocalitri commented:
Please rebase this PR

amir-qayyum-khan commented:
Done @giocalitri

amir-qayyum-khan commented:
jenkins run python
jenkins run lettuce

amir-qayyum-khan commented:
jenkins run python

doctoryes commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -433,8 +439,11 @@ def _accessible_courses_list_from_groups(request):
-                # ignore deleted or errored courses
+
+            # Custom Courses for edX (CCX) is an edX feature for re-using course content.
+#           # CCXs cannot be edited in Studio (aka cms) and should not be shown in this dashboard.
```
Nit: Extra `#`.

amir-qayyum-khan replied:
fixed

doctoryes commented on lms/djangoapps/ccx/utils.py:
```diff
@@ -260,3 +261,83 @@ def is_email(identifier):
+    list_staff = list_with_level(course, 'staff')
+    list_instructor = list_with_level(course, 'instructor')
+
+    with ccx_course(ccx_key) as course:
```
Perhaps this pattern is already spread throughout the CCX code - but this code is confusing to someone not familiar with CCX. What's the difference between the `course` parameter passed-in and the `course` that's retrieved via the context manager? If they are truly different, is there a different naming convention that could be used here?

amir-qayyum-khan replied:
@doctoryes  these two are one same course.
```with ccx_course(ccx_key) as course:``` means overridden course in which content of some fields is override. We access overridden course with special id call ``CCXLocator``.

@pdpinch correct me if i am wrong?

doctoryes replied:
@amir-qayyum-khan If the passed-in `course` and the `course` retrieved by `ccx_course(ccx_key)` are indeed the same, then why incur the extra modulestore call in `ccx_course()`? Why use the passed-in `course` for `list_with_level()` instead of just using the `course` gotten by `ccx_course()`? There's no explicit guarantee in this function that the passed-in `course` and the `course` retrieved by `ccx_course()` are indeed the same course.

amir-qayyum-khan replied:
@doctoryes  renamed there variables , please see new commit ,
thanks

doctoryes commented on lms/djangoapps/courseware/access.py:
```diff
@@ -132,9 +132,6 @@ def has_access(user, action, obj, course_key=None):
 
-    if isinstance(course_key, CCXLocator):
-        course_key = course_key.to_course_locator()
-
```
What's the net effect of not converting `CCXLocator`s to `CourseLocator`s here?

amir-qayyum-khan replied:
@doctoryes in courseware app the masquerading code which we use for viewing courseware content as staff, student , specific student, it use same cache which it stores against course_key/ccx key. So when we use this line ```course_key = course_key.to_course_locator()``` it returns master course key instead of ccx key but masquerade code needs ccx (for viewing ccx courseware as staff, student or specific student)

doctoryes replied:
Makes sense - so this change is a bug fix.

amir-qayyum-khan replied:
yes,

doctoryes commented on lms/djangoapps/ccx/views.py:
```diff
@@ -146,6 +147,9 @@ def dashboard(request, course, ccx=None):
             'ccx_set_grading_policy', kwargs={'course_id': ccx_locator})
 
+        with ccx_course(ccx_locator) as course:
+            context['course'] = course
```
This code adds `'course'` to the context - but I don't see any accompanying changes to the template, nor any reference in the existing template? Why was this attribute added to the context?

amir-qayyum-khan replied:
@doctoryes please see these lines
- https://github.com/edx/edx-platform/blob/master/lms/templates/ccx/coach_dashboard.html#L18
- https://github.com/edx/edx-platform/blob/master/lms/templates/courseware/course_navigation.html#L27-L33

doctoryes replied:
Thanks - I see it now.

This view uses the passed-in `course` to find the CCX - and then uses the found CCX's locator to retrieve the course again. Why not just use the passed-in course and skip the `ccx_course()` call?

amir-qayyum-khan commented:
jenkins run js

doctoryes commented on lms/djangoapps/ccx/tests/test_utils.py:
```diff
@@ -47,3 +61,74 @@ def test_ccx_locator(self):
+
+        # Create instructor account
+        self.coach = coach = AdminFactory.create()
+        self.client.login(username=coach.username, password="test")
```
Nit: Remove the extra `coach` and change to `username=self.coach.username`.

doctoryes commented:
My first pass is complete.

@edx/devops Please review the data migration in this PR.

pdpinch commented:
@amir-qayyum-khan can you add a comment(s) telling us what you've changed most recently -- or, in the future, you can add another commit so the reviewers can do a diff. If you go that route, you'd have to squash before the merge.

amir-qayyum-khan commented:
@pdpinch 

Oh, I squashed commits after this message https://github.com/edx/edx-platform/pull/10877#issuecomment-184751615,

- I renamed few variables in this file https://github.com/edx/edx-platform/pull/10877/files#diff-060752b618f1de90b37518736772a2cbR266
- Updated this test suit https://github.com/edx/edx-platform/pull/10877/files#diff-c309dd6446493eb28b2a3392667b55c9R65 , removed unnecessary code by incorporating  ``CcxTestCase``

jibsheet commented:
:+1: on the migration from DevOps

pdpinch commented:
:+1: from me. 

@doctoryes and @giocalitri did Amir address all your concerns?

doctoryes commented:
Yes - :+1:

giocalitri commented:
:+1: also from me
